DIPs/DIP27.md

```diff
@@ -0,0 +1,172 @@
+
+foo.bar(); // function bar
+
+void buzz(uint a) {}
+foo.buzz(); // foo called !
+```
+
+This effectively will allow the use of optional parenthesis on first class functions, which obviate the need to keep function as non first class object to be able to benefit from optional parenthesis.
+
+### Rationale
+
+D is a very complex languages. It promotes several paradigms while being a system language. If this is a force of D, this is also a source of trouble. In fact every feature can combine itself why many other feature, some of them from different paradigms - other way of thinking - and every single special case, cause a combinatorial explosion of special cases when other features are involved.
+
+This problem is experienced by many D users who try to use advanced feature together in a non trivial project. Combined together, special cases lead to surprising behaviors, unnecessary blockades, and sometime compiler bugs.
+
+If it is impairing users, it is also impairing polishing the language, as many corner cases have to be considered, and cause difficulties for compiler implementers and libraries creators that uses generic techniques.
+
+To solve, or at least reduce that problem, D specification needs to promote simplicity (as opposed to complexity, not difficulty). To achieve this goal, this DIP reduce all function to a single entity : the D first class function. It get rid of functions like defined in C or C++, as they are useless and essentially accidental complexity.
+
+### Breaking changes / deprecation process
+
+The unary & operator is defined as a NOOP when used on identifiers that resolve as function declarations. As this behavior clashes with the address of behavior, and is a special case, it may disappear after proper deprecation process is followed. Alternatively, it may be kept for backward compatibility.
+
+```d
+void foo() {}
+
+static assert(is(typeof(foo) == void function()); // Pass
+static assert(is(typeof(&foo) == void function()); // Transitional behavior.
+static assert(is(typeof(&foo) == void function()); // Error (foo has no address). Optional final behavior.
+static assert(is(typeof(&foo) == void function()*); // Pass with variant behavior expressed in possible variation section.
+```  
+
+The area where this DIP breaks code is when it comes to optionnal parentheses. The specification is such as most current uses of optional parenthesis will keep the same semantic. The current semantic is essentially accidental and a great source of confusion.
+
+Breakage will occur when a function's result is passed as argument to another function using optional parenthesis. The case is uncommon enough and even proved itself to be a confusing enough to be justification to introduce breaking changes in phobos, for instance, disabling argumentless text calls.
+
+The compiler should warn about this for a while before actually changing the behavior.
+
+## Copyright & License
+
+Copyright (c) 2016 by Amaury SÉCHET
```

Comment:
@Dicebot what is the policy on the copyright for new DIP's?
Is there a way this could hurt us or is there a potentialadvantage of licensing to the D Foundation?
(I am no expert by any means in this topic.)